content/courses/intro-to-embedded-sys-prog/chapters/handling_interrupts.rst

```diff
@@ -6,177 +6,1537 @@ Handling Interrupts
+semantics for interrupt handlers are as if a hardware "task," executing
+at an interrupt level priority, calls the protected procedure handler.
+
+Interrupt handlers in Ada are protected procedures, which do not have
+priorities individually, but the enclosing protected object can be
+assigned a priority, which will apply to the handler(s) when executing.
+
+Therefore, protected objects can have priorities assigned using values
+of subtype :ada:`System.Interrupt_Priority`, which are high enough to
+require the blocking of one or more interrupts. The specific values
+among the priority subtypes are not standardized but the intent is that
+interrupt priorities are higher (more urgent) than non-interrupt
+priorities, as if they are declared like so in package :ada:`System`:
+
+.. code-block:: ada
+
+   subtype Any_Priority is Integer range implementation-defined;
+
+   subtype Priority is Any_Priority
+      range Any_Priority'First .. implementation-defined;
+
+   subtype Interrupt_Priority is Any_Priority
+      range Priority'Last + 1 .. Any_Priority'Last;
+
+For example, here are the subtype declarations in the GNAT
+implementation for an Arm Cortex M4 target:
+
+.. code-block:: ada
+
+   subtype Any_Priority       is Integer      range 0 .. 255;
+   subtype Priority           is Any_Priority range Any_Priority'First .. 240;
+   subtype Interrupt_Priority is Any_Priority range Priority'Last + 1  .. Any_Priority'Last;
+
+Although the ranges are implementation-defined, when the Systems
+Programming Annex is implemented the range of
+:ada:`System.Interrupt_Priority` must include at least one value.
+Vendors are not required to have a distinct priority value in
+:ada:`Interrupt_Priority` for each hardware interrupt possible on a
+given target. On a bare-metal target, they probably will have a
+one-to-one correspondence, but might not in a target with an RTOS or
+host OS.
+
+A PO containing an interrupt handler procedure must be given a priority
+within the :ada:`Interrupt_Priority` subtype's range. To do so, we apply
+the aspect :ada:`Interrupt_Priority` to the PO. Perhaps confusingly, the
+aspect and the value's required subtype have the same name.
+
+.. code-block:: ada
+
+   with Ada.Interrupts.Names;  use Ada.Interrupts.Names;
+   with System;                use System;
+
+   package Gyro_Interrupts is
+
+      protected Handler with
+         Interrupt_Priority => Interrupt_Priority'Last
+      is
+      private
+         procedure IRQ_Handler;
+         pragma Attach_Handler (IRQ_Handler, EXTI2_Interrupt);
+      end Handler;
 
-    Complete section!
+   end Gyro_Interrupts;
 
+The code above uses the highest (most urgent) interrupt priority value
+but some other value could be used instead, as long as it is in the
+:ada:`Interrupt_Priority` subtype's range. :ada:`Constraint_Error` is
+raised otherwise.
 
-Implicit Elaboration Model
---------------------------
+There is also an alternative pragma, now obsolescent, with the same name
```

Comment:
Since this pragma is obsolescent, we could move the whole paragraph to an admonition, e.g.:

```
.. admonition:: Relevant topics

    There is also an alternative pragma, now obsolescent, ...
```